components/ws-daemon/nsinsider/main.go

```diff
@@ -155,6 +156,89 @@ func main() {
 					return nil
 				},
 			},
+			{
+				Name:  "mount-idmapped-mark",
+				Usage: "mounts a idmapped mark",
+				Flags: []cli.Flag{
+					&cli.StringFlag{
+						Name:     "source",
+						Required: true,
+					},
+					&cli.StringFlag{
+						Name:     "merged",
+						Required: true,
+					},
+					&cli.StringFlag{
+						Name:     "upper",
+						Required: true,
+					},
+					&cli.StringFlag{
+						Name:     "work",
+						Required: true,
+					},
+					&cli.StringFlag{
+						Name:     "userns",
+						Required: true,
+					},
+				},
+				Action: func(c *cli.Context) error {
+					target := filepath.Clean(c.String("merged"))
+					upper := filepath.Clean(c.String("upper"))
+					work := filepath.Clean(c.String("work"))
+					source := filepath.Clean(c.String("source"))
+					userns := filepath.Clean(c.String("userns"))
+
+					mappedLower, err := os.MkdirTemp("", "idmapped-mark")
+					if err != nil {
+						return xerrors.Errorf("mkdirtemp: %w", err)
+					}
+
+					mappedFD, err := unix.OpenTree(unix.AT_FDCWD, source, unix.OPEN_TREE_CLONE)
+					if err != nil {
+						return xerrors.Errorf("opentree: %w", err)
+					}
+					var closeOnce sync.Once
+					closeMappedFD := func() {
+						closeOnce.Do(func() { unix.Close(mappedFD) })
+					}
+					defer closeMappedFD()
+
+					usernsFD, err := os.Open(userns)
+					if err != nil {
+						return xerrors.Errorf("open(userns): %w", err)
+					}
+					defer usernsFD.Close()
+					err = unix.MountSetattr(mappedFD, "", unix.AT_EMPTY_PATH|unix.AT_RECURSIVE, &unix.MountAttr{
```

Comment:
@csweichel 
How about this one? Maybe it could work without EINVAL because the id-mapped mount doesn't need to mount the overlays layers. Or did it have to be recursive?
```suggestion
					err = unix.MountSetattr(mappedFD, "", unix.AT_EMPTY_PATH, &unix.MountAttr{
```

Reply:
Roger that.